src/Interpreters/SyntaxAnalyzer.cpp

```diff
@@ -529,6 +530,45 @@ void optimizeDuplicateOrderByAndDistinct(ASTPtr & query, const Context & context
     DuplicateDistinctVisitor(distinct_data).visit(query);
 }
 
+/// Optimize monotonous functions in ORDER BY
+void optimizeMonotonousFunctionsInOrderBy(ASTSelectQuery * select_query, const Context & context,
+                                          const TablesWithColumns & tables_with_columns)
+{
+    auto order_by = select_query->orderBy();
+    if (!order_by)
+        return;
+
+    std::unordered_map<String, ASTPtr> group_by_hashes;
```

Comment:
Why unordered_set is not enough? Also we can use `std::unordered_*<UInt128, ASTPtr, UInt128Hash>` directly.